### DISCUSSION TITLE
eBPF based TOA

### DISCUSSION THREAD

**Discussion by byxzone** [iovisor/bcc#4441] [General]
In ` BPF_PROG_TYPE_CGROUP_SOCK_ADDR ` program type, the `struct sock` cannot be accessed. I want to implement an eBPF-based TOA (Tcp Option Address) to restore the real client IP/PORT in LVS-FULLNAT mode, using `BPF_CGROUP_INET4_GETPEERNAME` attach mode

**Comment by chenhengqi** [comment]
Not familiar with BPF_PROG_TYPE_CGROUP_SOCK_ADDR , but you may use tc-bpf to implement TOA.

**Reply by byxzone** [reply]
In my impression, TC is before conntrack. Will this lead to the failure of fullnat?

**Reply by chenhengqi** [reply]
tc-bpf is pretty generic, it should be trivial to implement TOA or full NAT.
